File: ai-service/app/models/graph.py
from typing import Literal

from pydantic import BaseModel, Field

NodeType = Literal["STRATUM", "ROCK_BODY", "STRUCTURE", "ORE_BODY", "MINERAL", "REGION"]
RelationType = Literal["LOCATED_IN", "CONTAINS", "CONTROLS", "INTRUDES"]


class GraphNode(BaseModel):
    entity_id: int
    document_id: int
    name: str = Field(min_length=1, max_length=255)
    node_type: NodeType
    source_text: str = Field(min_length=1, max_length=2000)
    page: int = Field(ge=1)
    longitude: float | None = Field(default=None, ge=-180, le=180)
    latitude: float | None = Field(default=None, ge=-90, le=90)


class GraphRelation(BaseModel):
    source_entity_id: int
    target_entity_id: int
    relation_type: RelationType
    confidence: float = Field(ge=0, le=1)
    source_text: str = Field(min_length=1, max_length=2000)
    page: int = Field(ge=1)


class VectorChunk(BaseModel):
    chunk_id: int
    document_id: int
    document_name: str
    content: str = Field(min_length=1, max_length=20000)
    page_start: int = Field(ge=1)
    page_end: int = Field(ge=1)


class GraphSyncRequest(BaseModel):
    document_id: int
    nodes: list[GraphNode]
    relations: list[GraphRelation]
    chunks: list[VectorChunk]


class GraphSyncResponse(BaseModel):
    node_count: int
    relation_count: int
    vector_count: int


class GraphView(BaseModel):
    nodes: list[dict]
    links: list[dict]


class QuestionRequest(BaseModel):
    question: str = Field(min_length=2, max_length=1000)
    provider: Literal["deepseek", "qwen"] | None = None
    limit: int = Field(default=5, ge=1, le=12)


class QuestionResponse(BaseModel):
    answer: str
    related_entities: list[dict]
    spatial_locations: list[dict]
    sources: list[dict]
    provider: str
    model: str

File: frontend/src/api/graph.ts
import { http } from './http'

export type GraphNodeType='STRATUM'|'ROCK_BODY'|'STRUCTURE'|'ORE_BODY'|'MINERAL'|'REGION'
export interface GraphNode{id:number;name:string;nodeType:GraphNodeType;documentId:number;sourceText:string;page:number;longitude?:number;latitude?:number}
export interface GraphLink{source:number;target:number;relationType:'LOCATED_IN'|'OCCURS_IN'|'INTRUDES'|'CONTACTS'|'CONTROLS'|'CONTAINS';confidence?:number;sourceText?:string;page?:number}
export interface GraphView{nodes:GraphNode[];links:GraphLink[]}
export interface GraphStatus{documentId:number;status:'PENDING'|'SYNCING'|'COMPLETED'|'FAILED';progress:number;errorMessage?:string;nodeCount:number;relationCount:number;vectorCount:number;syncedAt?:string}
export interface QaResponse{answer:string;relatedEntities:Array<{id:number;name:string;nodeType:GraphNodeType;page:number}>;spatialLocations:Array<{entityId:number;name:string;longitude:number;latitude:number}>;sources:Array<{documentId:number;documentName:string;chunkId:number;pageStart:number;pageEnd:number;content:string;score:number}>;provider:string;model:string}
export interface QaStreamHandlers{onStatus?:(message:string)=>void;onMetadata?:(metadata:Omit<QaResponse,'answer'>)=>void;onDraft?:(content:string)=>void;onReset?:()=>void;onDelta?:(content:string)=>void;onWarning?:(message:string)=>void}

export async function syncGraph(documentId:number):Promise<GraphStatus>{const response=await http.post(`/documents/${documentId}/graph/sync`,undefined,{timeout:120000});return response.data.data}
export async function getGraphStatus(documentId:number):Promise<GraphStatus>{const response=await http.get(`/documents/${documentId}/graph/status`);return response.data.data}
export async function queryGraphNodes(query='',documentId?:number):Promise<GraphView>{const response=await http.get('/graph/nodes',{params:{query,documentId,limit:200}});return response.data.data}
export async function expandGraphNode(id:number,depth=1):Promise<GraphView>{const response=await http.get(`/graph/nodes/${id}/expand`,{params:{depth}});return response.data.data}
export async function queryGraphPath(sourceId:number,targetId:number):Promise<GraphView>{const response=await http.get('/graph/path',{params:{sourceId,targetId}});return response.data.data}
export async function askGeologicalQuestion(question:string,provider:'deepseek'|'qwen',limit=5):Promise<QaResponse>{const response=await http.post('/qa/ask',{question,provider,limit},{timeout:180000});return response.data.data}
export async function streamGeologicalQuestion(question:string,provider:'deepseek'|'qwen',handlers:QaStreamHandlers,limit=5,signal?:AbortSignal):Promise<void>{
  const token=localStorage.getItem('geotext-token')
  const response=await fetch('/api/qa/ask/stream',{method:'POST',headers:{'Content-Type':'application/json',Accept:'text/event-stream',...(token?{Authorization:`Bearer ${token}`}:{})},body:JSON.stringify({question,provider,limit}),signal})
  if(response.status===401){localStorage.removeItem('geotext-token');localStorage.removeItem('geotext-profile');throw new Error('登录状态已失效，请重新登录')}
  if(!response.ok||!response.body)throw new Error(`智能问答服务返回 HTTP ${response.status}`)
  const reader=response.body.getReader(),decoder=new TextDecoder();let buffer=''
  const dispatch=(frame:string)=>{let event='message';const data:string[]=[];for(const line of frame.split(/\r?\n/)){if(line.startsWith('event:'))event=line.slice(6).trim();else if(line.startsWith('data:'))data.push(line.slice(5).trim())}if(!data.length)return;const payload=JSON.parse(data.join('\n'));if(event==='status')handlers.onStatus?.(payload.message);else if(event==='metadata')handlers.onMetadata?.({relatedEntities:payload.related_entities,spatialLocations:payload.spatial_locations,sources:payload.sources,provider:payload.provider,model:payload.model});else if(event==='draft')handlers.onDraft?.(payload.content);else if(event==='reset')handlers.onReset?.();else if(event==='delta')handlers.onDelta?.(payload.content);else if(event==='warning')handlers.onWarning?.(payload.message);else if(event==='error')throw new Error(payload.message||'智能问答失败')}
  while(true){const {done,value}=await reader.read();buffer+=decoder.decode(value||new Uint8Array(),{stream:!done});const frames=buffer.split(/\r?\n\r?\n/);buffer=frames.pop()||'';for(const frame of frames)dispatch(frame);if(done)break}
  if(buffer.trim())dispatch(buffer)
}
